File: 204 Count Primes.c
#include <stdio.h>
#include <stdlib.h>
/*
Count Primes
https://leetcode.com/problems/count-primes/

Regular Eratosthenes method
*/
int eratosthenes(int x){
    int i,j;
    char* bc = (char*)calloc(x+1,sizeof(char)); // 0 means prime
    int counter = 0;
    
    bc[0] = 1;
    bc[1] = 1;
    for(i = 2;i<=x;i++)
        if(bc[i]==0){
            counter++;
            for(j = i*i;j<=x;j += i)
                bc[j]=1;
        }
    free(bc);
    return counter;
}
int main()
{
    // 100 -> 25
    printf("%d\n",eratosthenes(100));
    return 0;
}

File: 204 Count Primes.py
'''
204. Count Primes
https://leetcode.com/problems/count-primes/

Count the number of prime numbers less than a non-negative number, n.

Example:

Input: 10
Output: 4
Explanation: There are 4 prime numbers less than 10, they are 2, 3, 5, 7.
'''
class Solution:
    def countPrimes(self, n: int) -> int:
        '''
        Eratosthenes
        '''
        result = 0
        
        # primes[i] is prime if True
        primes = [True] * n
        
        for i in range(2, n):
            if primes[i]:
                result += 1
                
                # Mark all multiple as non-prime
                m = i
                while m * i < n:
                    primes[m * i] = False
                    m += 1

        return result
